plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java

```diff
@@ -789,12 +805,85 @@ public WriteMapping toWriteMapping(ConnectorSession session, Type type)
         }
+        if (type instanceof TimestampType timestampType) {
+            return timestampWriteMapping(timestampType);
+        }
+        if (type instanceof TimestampWithTimeZoneType timestampWithTimeZoneType) {
+            return timestampWithTimeZoneWriteMapping(session.getTimeZoneKey(), timestampWithTimeZoneType);
+        }
         if (type.equals(uuidType)) {
             return WriteMapping.sliceMapping("UUID", uuidWriteFunction());
         }
         throw new TrinoException(NOT_SUPPORTED, "Unsupported column type: " + type);
     }
 
+    private WriteMapping timestampWriteMapping(TimestampType type)
+    {
+        int precision = type.getPrecision();
+        String dataType = "DateTime64(%s)".formatted(precision);
+        if (type.isShort()) {
+            return WriteMapping.longMapping(dataType, timestampWriteFunction(createTimestampType(precision)));
+        }
+        checkArgument(precision <= CLICKHOUSE_MAX_SUPPORTED_TIMESTAMP_PRECISION, "Precision is out of range: %s", precision);
+        return WriteMapping.objectMapping(dataType, longTimestampWriteFunction(type, precision));
+    }
+
+    private static ColumnMapping timestampWithTimeZoneColumnMapping(ClickHouseColumn clickHouseColumn)
+    {
+        int precision = clickHouseColumn.getScale();
+        TimeZone columnTimeZone = clickHouseColumn.getTimeZone();
+        checkArgument(precision <= CLICKHOUSE_MAX_SUPPORTED_TIMESTAMP_PRECISION, "Precision is out of range: %s", precision);
+        TimestampWithTimeZoneType trinoType = createTimestampWithTimeZoneType(precision);
+        if (trinoType.isShort()) {
+            return ColumnMapping.longMapping(
+                    trinoType,
+                    shortTimestampWithTimeZoneReadFunction(),
+                    shortTimestampWithTimeZoneWriteFunction(columnTimeZone));
+        }
+        return ColumnMapping.objectMapping(
+                trinoType,
+                longTimestampWithTimeZoneReadFunction(),
+                longTimestampWithTimeZoneWriteFunction(columnTimeZone));
+    }
+
+    private static WriteMapping timestampWithTimeZoneWriteMapping(TimeZoneKey timeZoneKey, TimestampWithTimeZoneType timestampWithTimeZoneType)
+    {
+        verify(timestampWithTimeZoneType.getPrecision() <= CLICKHOUSE_MAX_SUPPORTED_TIMESTAMP_PRECISION);
+        // Clickhouse DateTime64(precision, [timezone])
+        // In Clickhouse the time zone is not stored in the rows of the table (or in resultset), but is stored in the column metadata.
+        // Timezone agnostic Unix timestamp is stored in tables
+        // In trino, timezone is not available at the point of time when write mapping is resolved
+        // So to allow to create columns within different timezones within Clickhouse - use timezone from trino session.
+        TimeZone timeZone = TimeZone.getTimeZone(timeZoneKey.getZoneId());
```

Comment:
@Praveen2112 I've some doubts about obtaining zone from session `session.getTimeZoneKey()` implementation vs hardcoding it to 'UTC'

Reply:
Also, it relates to CI fails:

<details>
<summary>stacktraces</summary>

Error:  Tests run: 204, Failures: 2, Errors: 0, Skipped: 78, Time elapsed: 285.2 s <<< FAILURE! -- in io.trino.plugin.clickhouse.TestClickHouseConnectorTest
Error:  io.trino.plugin.clickhouse.TestClickHouseConnectorTest.testTimestampWithTimeZoneCastToTimestampPredicate -- Time elapsed: 0.712 s <<< FAILURE!
org.assertj.core.error.AssertJMultipleFailuresError: 

Multiple Failures (1 failure)
2024-11-18T07:15:25.402-0600	INFO	main	io.trino.testing.containers.junit.ReportLeakedContainers	Checking for leaked containers
-- failure 1 --
[Rows for query [SELECT i FROM timestamptz_to_ts4uhjlttnkq WHERE CAST(t AS timestamp(0)) = TIMESTAMP '2005-09-10 13:00:00']] 
Expecting RegularImmutableList:
  
to contain:
  [(UTC)]
but could not find the following element(s):
  [(UTC)]
at QueryAssertions$ResultAssert.lambda$containsAll$1(QueryAssertions$ResultAssert.java:756)
	at io.trino.sql.query.QueryAssertions$ResultAssert.containsAll(QueryAssertions.java:748)
	at io.trino.sql.query.QueryAssertions$ResultAssert.containsAll(QueryAssertions.java:721)
	at io.trino.sql.query.QueryAssertions$QueryAssert.containsAll(QueryAssertions.java:382)
	at io.trino.testing.BaseConnectorTest.testTimestampWithTimeZoneCastToTimestampPredicate(BaseConnectorTest.java:6018)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1458)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2034)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:189)

Error:  io.trino.plugin.clickhouse.TestClickHouseConnectorTest.testTimestampWithTimeZoneCastToDatePredicate -- Time elapsed: 0.831 s <<< FAILURE!
org.assertj.core.error.AssertJMultipleFailuresError: 

Multiple Failures (1 failure)
-- failure 1 --
[Rows for query [SELECT i FROM timestamptz_to_date65vvo66juq WHERE CAST(t AS date) = DATE '2005-09-10']] 
Expecting RegularImmutableList:
  
to contain:
  [(UTC), (Los Angeles)]
but could not find the following element(s):
  [(UTC), (Los Angeles)]
at QueryAssertions$ResultAssert.lambda$containsAll$1(QueryAssertions$ResultAssert.java:756)
	at io.trino.sql.query.QueryAssertions$ResultAssert.containsAll(QueryAssertions.java:748)
	at io.trino.sql.query.QueryAssertions$ResultAssert.containsAll(QueryAssertions.java:721)
	at io.trino.sql.query.QueryAssertions$QueryAssert.containsAll(QueryAssertions.java:382)
	at io.trino.testing.BaseConnectorTest.testTimestampWithTimeZoneCastToDatePredicate(BaseConnectorTest.java:5987)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1458)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2034)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:189)

</details>

Reply:
Instead of fetching them from Session or hardcoding it to `UTC` - What if we could capture them as a column property ? This would allow us to customize it for each column as they require. WDYT ?

Reply:
It will not be able to support properties for CTAS. 
So hardcoding it to 'UTC' looks like a good enough solution.

Reply:
As agreed, not allowing CREATE TABLE or CTAS for now.